dotnet-statsig/src/Statsig/Server/Evaluation/SpecStore.cs

```diff
@@ -24,104 +25,199 @@ internal IDList(string name)
 
     class SpecStore
     {
-        double _lastSyncTime;
-        RequestDispatcher _requestDispatcher;
-        Timer _syncTimer;
-        Timer _idListSyncTimer;
+        private double _lastSyncTime;
+        private readonly RequestDispatcher _requestDispatcher;
 
-        internal Dictionary<string, ConfigSpec> FeatureGates { get; set; }
-        internal Dictionary<string, ConfigSpec> DynamicConfigs { get; set; }
-        internal Dictionary<string, IDList> IDLists { get; set; }
+        private Task _syncIDListsTask;
+        private Task _syncValuesTask;
+        private readonly CancellationTokenSource _cts;
+
+        internal Dictionary<string, ConfigSpec> FeatureGates { get; private set; }
+        internal Dictionary<string, ConfigSpec> DynamicConfigs { get; private set; }
+        private readonly Dictionary<string, IDList> _idLists;
+        private readonly ReaderWriterLockSlim _idListLock;
 
         internal SpecStore(string serverSecret, StatsigOptions options)
         {
             _requestDispatcher = new RequestDispatcher(serverSecret, options.ApiUrlBase);
             _lastSyncTime = 0;
             FeatureGates = new Dictionary<string, ConfigSpec>();
             DynamicConfigs = new Dictionary<string, ConfigSpec>();
-            IDLists = new Dictionary<string, IDList>();
+            _idLists = new Dictionary<string, IDList>();
+            _idListLock = new ReaderWriterLockSlim();
+
+            _syncIDListsTask = null;
+            _syncValuesTask = null;
+            _cts = new CancellationTokenSource();
         }
 
         internal async Task Initialize()
         {
+            // Execute and wait for the initial syncs
             await SyncValues(true);
             await SyncIDLists();
+
+            // Start background tasks to periodically refresh the store
+            _syncValuesTask = BackgroundPeriodicSyncValuesTask(_cts.Token);
+            _syncIDListsTask = BackgroundPeriodicSyncIDListsTask(_cts.Token);
         }
 
-        internal void Shutdown()
+        internal async Task Shutdown()
         {
-            _syncTimer.Stop();
-            _syncTimer.Dispose();
-            _idListSyncTimer.Stop();
-            _idListSyncTimer.Dispose();
+            // Signal that the periodic task should exit, and then wait for them to finish
+            _cts.Cancel();
+            if (_syncIDListsTask != null)
+            {
+                await _syncIDListsTask;
+            }
+            if (_syncValuesTask != null)
+            {
+                await _syncValuesTask;
+            }
```

Comment:
we don't need to wait for them finish as they are only needed if the server is running to serve the most up to date values. Can we shorten the time needed to shutdown by not waiting for them, or some other alternative if cleanup is needed to prevent certain bad things?

Reply:
In general these tasks should abort virtually instantly when the cancellation token is signalled as the tasks should be in the middle of a Task.Delay().  The only case in which the task wouldn't abort immediately would be if it was in the middle of executing the HTTP calls for one of the periodic syncs.  As part of a follow-up PR, I was going to address the HTTP layer and part of that would involve plumbing the CancellationTokens through to that code so that it too can be quickly aborted as necessary.

Reply:
Makes sense!